src/background/index.ts

```diff
@@ -46,23 +44,40 @@ browser.webRequest.onBeforeSendHeaders.addListener(
 )
 
 // Update on tab change or new urls
-browser.tabs.onActivated.addListener(() => getCurrentLocale())
-browser.tabs.onUpdated.addListener(() => getCurrentLocale())
+browser.tabs.onActivated.addListener((tabId) => setBadge(tabId.tabId))
+browser.tabs.onUpdated.addListener(async () => {
+  const tabId = await getCurrentTabId()
+  if (tabId) setBadge(tabId)
+})
+
+function sendCurrentState(tabId: number) {
+  const locale: Locale = (tabId && ram.get(tabId)) || null
+  const msg: MessageType = { type: 'current-output', data: locale }
+  browser.runtime.sendMessage(msg)
+}
 
 // Listener for messages from content scripts
-browser.runtime.onMessage.addListener(({ type, data }) => {
-  switch (type) {
-    case 'get':
-      updateLocale(data)
+browser.runtime.onMessage.addListener(async (message: MessageType, sender) => {
+  switch (message.type) {
+    case 'sync': {
+      const tabId = sender.tab?.id ?? null
+      if (tabId) updateLocale(tabId, message.data)
       break
-  }
-})
+    }
 
-// Listen for changes to the locale
-browser.storage.onChanged.addListener((changes) => {
-  const change = changes['locale']
-  if (change) {
-    updateLocale(change.newValue)
-    sendMessageToActiveTabs({ type: 'update', data: locale })
+    case 'update': {
+      const tabId = await getCurrentTabId()
+      if (tabId) {
+        updateLocale(tabId, message.data)
+        const msg: MessageType = { type: 'update', data: message.data }
+        browser.tabs.sendMessage(tabId, msg)
+      }
+      break
+    }
+
+    case 'current-input': {
+      const tabId = await getCurrentTabId()
+      if (tabId) sendCurrentState(tabId)
```

Comment:
I think we can actually respond immediately using `sendResponse`, rather than broadcast another event: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sending_a_synchronous_response

Reply:
True, but we needed an event anyways since we need to listen for changes, so I thought i'll keep it consistent